### ISSUE TITLE
get allDocs as stream

### ISSUE THREAD

**Issue by mathanpt** [cloudant/java-cloudant#344]
java-cloudant version - 2.7.0
Java version - 1.8

code:
`Database dbs = client.database("test", false);
    AllDocsRequest allDocsRequest = dbs.getAllDocsRequestBuilder().includeDocs(true).build();
    InputStream inputStream = (InputStream) allDocsRequest.getResponse().getDocs();`

how to get allDocs as Stream?

**Comment by ricellis** [comment]
What is your use case for using the stream directly as opposed to de-serializing the docs with [`allDocsRequest.getResponse().getDocsAs(YourDocType.class)`](http://static.javadoc.io/com.cloudant/cloudant-client/2.7.0/com/cloudant/client/api/views/AllDocsResponse.html#getDocsAs-java.lang.Class-)? Is it because you use your own non-GSON deserializer? I'm interested in understanding how the library is being used because we have at times considered, for example, providing plug points for user-provided deserializers, or a per-document callback interface for processing streams that include documents.

There is no API for accessing the `InputStream` directly from the `AllDocsResponse`, but it is possible to get an `InputStream` by making the request using the [client.executeRequest](http://static.javadoc.io/com.cloudant/cloudant-client/2.7.0/com/cloudant/client/api/CloudantClient.html#executeRequest-com.cloudant.http.HttpConnection-). For this case that would look like:

```java
Database dbs = client.database("test", false);
HttpConnection response = client.executeRequest(Http.GET(
           new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true")));
// Use try-with-resources or try/finally to ensure stream is closed after use
try (InputStream inputStream = response.responseAsInputStream()) {
     // process stream
}
```

**Comment by mathanpt** [comment]
@ricellis Thanks for the help I am able to get the InputStream

Is there any way to fetch all_docs except design documents I have tried with **startKey** and **endKey** as query params but didn't work

**Comment by ricellis** [comment]
@mathanpt 
I think using `startkey` and `endkey` are the only way. You have to be careful though depending what your document IDs are like (e.g. if they are all numbers, always start lower case or use a mixture like the generated UUIDs). This is because  `_all_docs` follows [the ASCII collation sequence not unicode](http://docs.couchdb.org/en/2.0.0/couchapp/views/collation.html#all-docs) so the `_` is inbetween upper and lower case letters. This means you might need to do two requests to really get all of the docs, one to get everything before `_design` and then another for everything after.

For example, for the stream workaround:
```java
// Get all docs with IDs before _design (e.g. starting with numbers or uppercase letters)
HttpConnection response = client.executeRequest(Http.GET(
           new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true&endkey=\"_design\"")));

//Get all docs with IDs from _design0 (e.g. starting with lowercase letters)
HttpConnection response = client.executeRequest(Http.GET(
           new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true&startkey=\"_design0\"")));
```

And for reference for anyone using the normal `AllDocsRequestBuilder`:
```java
// Get all docs with IDs before _design (e.g. starting with numbers or uppercase letters)
dbs.getAllDocsRequestBuilder().includeDocs(true).endKey("_design").build()

//Get all docs with IDs from _design0 (e.g. starting with lowercase letters)
dbs.getAllDocsRequestBuilder().includeDocs(true).startKey("_design0").build();
```

**Comment by mathanpt** [comment]
Below is the code I am trying with

```
package com.palmtree.cloudant.sample;

import com.cloudant.client.api.ClientBuilder;
import com.cloudant.client.api.CloudantClient;
import com.cloudant.client.api.Database;
import com.cloudant.http.Http;
import com.cloudant.http.HttpConnection;
import org.apache.commons.io.IOUtils;

import java.io.*;
import java.net.URL;
import java.util.List;

public class cloudantSampleClient {
  public static void main(String args[]) throws IOException {
    CloudantClient client = ClientBuilder.account(test)
            .username(user)
            .password(password)
            .build();

    List<String> databases = client.getAllDbs();
    System.out.println("All my databases : ");
    for ( String db : databases ) {
      System.out.println(db);
    }

    Database dbs = client.database("test", false);

    HttpConnection response = client.executeRequest(Http.GET(
            new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true&endkey=" + "%22_design%22")));

    HttpConnection response1 = client.executeRequest(Http.GET(
            new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true&startkey=" + "%22_design0%22")));

    try (InputStream inputStream1 = response.responseAsInputStream()) {
      InputStream inputStream2 = response1.responseAsInputStream();
      FileOutputStream fos1 = new FileOutputStream(new File("/home/mt/tmp/result"));
      IOUtils.copy(inputStream1, fos1);

      FileOutputStream fos2 = new FileOutputStream(new File("/home/mt/tmp/result2"));
      IOUtils.copy(inputStream2, fos2);

      FileInputStream fin1 = new FileInputStream("/home/mt/tmp/result");
      FileInputStream fin2 = new FileInputStream("/home/mt/tmp/result2");
      FileOutputStream fos = new FileOutputStream("/home/mt/tmp/output");
      SequenceInputStream sis = new SequenceInputStream(fin1, fin2);
      int i;
      while ((i=sis.read())!=-1) {
        fos.write(i);
      }
      sis.close();
      fos.close();
      fin1.close();
      fin2.close();
    }
  }
}

```

but I got an SSL exception while writing stream to "result2" file

> Exception in thread "main" javax.net.ssl.SSLException: SSL peer shut down incorrectly
>         at sun.security.ssl.InputRecord.readV3Record(InputRecord.java:596)
>         at sun.security.ssl.InputRecord.read(InputRecord.java:532)
>         at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973)
>         at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:930)
>         at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
>         at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
>         at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
>         at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
>         at sun.net.www.http.ChunkedInputStream.fastRead(ChunkedInputStream.java:244)
>         at sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:689)
>         at java.io.FilterInputStream.read(FilterInputStream.java:133)
>         at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3336)
>         at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3329)
>         at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1792)
>         at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1769)
>         at org.apache.commons.io.IOUtils.copy(IOUtils.java:1744)
>         at com.palmtree.cloudant.sample.cloudantSampleClient.main(cloudantSampleClient.java:52)
>

**Comment by ricellis** [comment]
@mathanpt Your sample code works fine for me, although I was only reading a small number of documents.

The exception indicates an issue with termination of the SSL connection, which I guess could be down to some network issue if it was a one off. If it is repeatable then one possibility is that making the two requests at the same time, but reading the streams sequentially means that the connection for `response1` sits idle while you are processing `response`/`inputStream1`. The server will prune connections that are left idle for a while so it might be better to make the first request and process the stream and then make the second request (or process the two streams in parallel).

I'm not sure if it is having an impact, but I also notice that you have both stream reads within the `try` for `inputStream1` and you never close `inputStream2`. It might be worth ensuring that `inputStream1` is closed when you finish reading it instead of leaving it open.

You could make the requests in sequence and simplify your stream close handling into something like this:

```java
// Make first request and process first response
HttpConnection response = client.executeRequest(Http.GET(
            new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true&endkey=" + "%22_design%22")));

try (InputStream inputStream1 = response.responseAsInputStream();
     FileOutputStream fos1 = new FileOutputStream(new File("/home/mt/tmp/result"))) {
         IOUtils.copy(inputStream1, fos1);
}

// Make second request and process second response
HttpConnection response1 = client.executeRequest(Http.GET(
            new URL(dbs.getDBUri().toString() + "/_all_docs?include_docs=true&startkey=" + "%22_design0%22")));

try (InputStream inputStream2 = response1.responseAsInputStream();
     FileOutputStream fos2 = new FileOutputStream(new File("/home/mt/tmp/result2"))) {
         IOUtils.copy(inputStream2, fos2);
}

// Combine response files
try (FileInputStream fin1 = new FileInputStream("/home/mt/tmp/result");
     FileInputStream fin2 = new FileInputStream("/home/mt/tmp/result2");
     FileOutputStream fos = new FileOutputStream("/home/mt/tmp/output");
     SequenceInputStream sis = new SequenceInputStream(fin1, fin2)) {
         IOUtils.copy(sis, fos);
}
```

**Comment by mathanpt** [comment]
@ricellis Thanks a lot. Sequential connection works fine and I am able to get all_docs but whereas in the output file I have two json arrays since we make two separate calls to the cloudant each one of them returns total_rows

e.g
$ grep -r "total" *
output:{"total_rows":xxxxxx,"offset":0,"rows":[
output:{"total_rows":xxxxxx,"offset":yyyyy,"rows":[

Is there any simplest way to remove two json array and make it as one?

**Comment by ricellis** [comment]
The simplest way I can think of to write a single array of all the docs except design docs into a file using Java 8 is this:
```java
List<Map> docs = dbs.getAllDocsRequestBuilder().includeDocs(true).build().getResponse().getDocsAs(Map.class);
docs.removeIf(d -> ((String) d.get("_id")).startsWith("_design"));
try (FileWriter fw = new FileWriter("/home/mt/tmp/output")) {
    new Gson().toJson(docs, fw);
}
```
As you can see it is a lot more straightforward to use the `Collection` rather than using the stream directly (assuming that you are not reading so many documents that they wouldn't fit into memory). Additionally if you use the `Collection` you can avoid the need to make two requests because you can filter the design documents out of the `List`.

Since you never answered my question about your use case for using the stream directly it is hard to know whether that approach would be suitable for what you want.

If you definitely need to work with the stream directly then I think you'll probably want to use a streaming JSON API to do whatever transformations you need with the streams. Although if you are using a streaming API, again, it may be better to just filter the design documents out of the stream as you process it instead of making two requests.

**Comment by ricellis** [comment]
Closing as there has been no discussion for a while, please re-open or open a new issue if there are additional questions.
